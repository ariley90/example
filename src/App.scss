body {
  background-image: url("./background.png");
  background-size: 100% 100%;
  background-repeat: no-repeat;
  height: 100vh;
  background-color: white;
}

.App {
  margin: 30px auto;
  width: 80%;

  display: flex;
  &__sidebar {
    width: 25%;
    padding-right: 5%;
  }

  &__content {
    width: 75%;
  }
}

.Sort {
  list-style: none;
  padding-inline-start: 0;
  margin: 0;
  li {
    border-top: 1px solid #17317f;
    &:first-of-type {
      border-top: 0;
    }
    &:hover {
      cursor: pointer;
      // background-color: #8995bb;
    }
    button {
      color: #17317f;
      background-color: white;
      padding: 15px;
      width: 100%;
      height: 100%;
      border: 0;
      font-size: 1rem;
      text-align: start;
    }
  }
  .selected {
    background-color: #17317f;
    color: white;
  }
}

.Result {
  margin-bottom: 20px;
  &__row {
    display: flex;
    position: relative;
  }
  &__image {
    width: 60%;
    object-fit: cover;
  }

  &__info {
    padding: 15px;
    background-color: white;
    width: 40%;
    font-size: 0.8rem;

    button {
      cursor: pointer;
      font-weight: bold;
      background-color: #fedc07;
      color: #17317f;
      border: 0;
      border-radius: 5px;
      width: 100%;
      margin: 10px auto;
      padding: 5px;
      &:hover {
        background-color: #dfc208;
      }
    }
  }

  &__title {
    font-size: 1.4rem;
    margin: 10px 0;
    font-weight: bold;
    color: #17317f;
  }

  &__subTitle {
    font-size: 0.9rem;
    color: gray;
  }

  &__cta {
    font-size: 1.4rem;
    font-weight: bold;
  }

  &__overlay {
    position: absolute;
    bottom: 0;
    left: 0;
    background-color: white;
    color: #17317f;
    padding: 10px;
    cursor: pointer;
  }

  &__overview {
    background-color: white;
    color: #17317f;
  }
}
.ratings {
  margin: 10px 0;
}
.rating {
  height: 15px;
  width: 15px;
}
