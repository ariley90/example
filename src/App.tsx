import React, { useState } from "react";
import "./App.scss";
import data from "./data.json";
import Sort, { Option } from "./Sort";
import Result, { IResult } from "./Result";

export type SortFn = (a: IResult, b: IResult) => number;
const sortByName = (a: IResult, b: IResult) => a.name.localeCompare(b.name);
const sortByPrice = (a: IResult, b: IResult) => b.price - a.price;
const sortByRating = (a: IResult, b: IResult) => b.rating - a.rating;

const sortOptions: Option[] = [
  {
    name: (
      <>
        sort <strong>alphabetically</strong>
      </>
    ),
    sortFn: sortByName,
  },
  {
    name: (
      <>
        sort by <strong>price</strong>
      </>
    ),
    sortFn: sortByPrice,
  },
  {
    name: (
      <>
        sort by <strong>rating</strong>
      </>
    ),
    sortFn: sortByRating,
  },
];

function App() {
  const [sortBy, setSortBy] = useState<SortFn>(() => sortByPrice);

  return (
    <div className="App">
      <div className="App__sidebar">
        <Sort
          options={sortOptions.map((so) => {
            return {
              ...so,
              selected: so.sortFn === sortBy,
            };
          })}
          onClick={(option) => setSortBy(() => option.sortFn)}
        />
      </div>
      <div className="App__content">
        {data.sort(sortBy).map((holiday) => (
          <Result key={holiday.ID} {...holiday} />
        ))}
      </div>
    </div>
  );
}

export default App;
