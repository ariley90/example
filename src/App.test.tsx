import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "./App";

test("by default it should sort the results by price", () => {
  render(<App />);
  expect(screen.getByRole("list")).toMatchSnapshot();
  const results = screen.getAllByRole("heading");
  expect(results[0]).toHaveTextContent("Iberostar Grand Salome");
  expect(results[1]).toHaveTextContent("Aguamarina Golf Hotel");
  expect(results[2]).toHaveTextContent("Las Piramides Resort");
});

test("results can be sorted alphapetically", () => {
  render(<App />);
  userEvent.click(screen.getByRole("button", { name: "sort alphabetically" }));
  const results = screen.getAllByRole("heading");
  expect(results[0]).toHaveTextContent("Aguamarina Golf Hotel");
  expect(results[1]).toHaveTextContent("Iberostar Grand Salome");
  expect(results[2]).toHaveTextContent("Las Piramides Resort");
});

test("results can be sorted by rating", () => {
  render(<App />);
  userEvent.click(screen.getByRole("button", { name: "sort by rating" }));
  const results = screen.getAllByRole("heading");
  expect(results[0]).toHaveTextContent("Iberostar Grand Salome");
  expect(results[1]).toHaveTextContent("Las Piramides Resort");
  expect(results[2]).toHaveTextContent("Aguamarina Golf Hotel");
});

test("a search result can toggle an overview ", () => {
  render(<App />);
  expect(screen.queryByText("overview 1")).not.toBeInTheDocument();
  userEvent.click(screen.getAllByText(/Read more/i)[0]);
  expect(screen.getByText("overview 1")).toBeInTheDocument();
  userEvent.click(screen.getByText(/Read less/i));
  expect(screen.queryByText("overvi ew 1")).not.toBeInTheDocument();
});

test("the visibility of a result overview should be maintained when re-sorting results", () => {
  render(<App />);
  expect(screen.queryByText("overview 1")).not.toBeInTheDocument();
  userEvent.click(screen.getAllByText(/Read more/i)[0]);
  expect(screen.getByText("overview 1")).toBeInTheDocument();
  userEvent.click(screen.getByRole("button", { name: "sort by rating" }));
  expect(screen.getByText("overview 1")).toBeInTheDocument();
});
