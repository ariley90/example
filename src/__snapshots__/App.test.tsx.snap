// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`by default it should sort the results by price 1`] = `
<ul
  class="Sort"
>
  <li>
    <button
      class=""
    >
      sort 
      <strong>
        alphabetically
      </strong>
    </button>
  </li>
  <li>
    <button
      class="selected"
    >
      sort by 
      <strong>
        price
      </strong>
    </button>
  </li>
  <li>
    <button
      class=""
    >
      sort by 
      <strong>
        rating
      </strong>
    </button>
  </li>
</ul>
`;
